### ISSUE TITLE
W-series GPU dispatch silently falls back to CPU: helper zeros h->shader_* fields the dispatch FFI reads

### ISSUE THREAD

**Issue by johnjezl** [SLM-OS/SLM-Operating-System#832]
## Root cause

The W-series per-op dispatch FFIs in `kernel/src/slm_ffi.c` (`slm_runtime_dispatch_rmsnorm_simt`, `_swiglu_simt`, `_gqa_attn_simt`, `_q4k_dot_simt`, `_embedding_simt`) all read the v4-era single-shader fields from the inherited channel handoff:

```c
const struct ga10b_channel_handoff *h = ga10b_bringup_handoff();
if (h == NULL || h->shader_gpu_va == 0 || h->shader_phys == 0 ||
    h->shader_size < OPLIB_POOL_MIN_BYTES) {
    spin_unlock_irqrestore(&g_gpu_dispatch_lock, irq);
    return -1;
}
```

But the current MNIST helper (`scripts/gpu-launch-common.c::gpu_write_handoff_v6` and `gpu_write_handoff_v7`) **explicitly zero those fields**:

```c
/* v6 = v5 + input_buf_phys/size. v3 dispatch fields stay zero
 * (same defensive zeroing as v5 — forces the v5 pipeline path
 * if pipeline_n_ops survives but other fields don't). */
struct ga10b_channel_handoff hoff = {
    ...
    .shader_phys        = 0,
    .shader_gpu_va      = 0,
    .shader_size        = 0,
    ...
};
```

Result: every W-series dispatch FFI returns -1 on entry. The hybrid wrappers in `forward.rs` translate that to `BackendError::NotAvailable` and silently fall through to the CPU NEON path. The GPU does **zero useful work** during an `slm prompt` run on Jetson, despite the tier table reading SIMT for six ops and `oplib stage: ok` reporting success.

## Hardware evidence

Per the new `slm gpu` counters (PR #831):

```
op           tier  attempts          ok          fallback
-----------  ----  ----------------  ----------  --------
RMSNORM      SIMT  3538              0           3538
GQA_ATTN     SIMT  1740              0           1740
ROPE         SIMT  0                 0           0
EMBEDDING    SIMT  0                 0           0
Q4K_DOT      SIMT  0                 0           0
SWIGLU       SIMT  0                 0           0
```

For a single 58-token SmolLM2-135M `Hello` prompt:
- RMSNORM: 3538 attempts, all -1 (helper handoff `shader_*` = 0)
- GQA_ATTN: 1740 attempts, all -1 (same)
- ROPE / EMBEDDING / Q4K_DOT / SWIGLU: zero attempts (the hybrid wrappers gate on conditions that exclude these — `gpu_tensor_map` entry missing because no weights pool was allocated by the helper; SWIGLU per W7 commit message is "inert until the FFN refactor adopts the hybrid wrapper")

`oplib stage` verbose dump confirms:
```
oplib stage: handoff h=0x80640060 shader_phys=0x0 shader_gpu_va=0x0 shader_size=0 cbuf_phys=0x0 cbuf_gpu_va=0x0
```

The v6/v7 handoff write paths zero these intentionally, and there's no in-tree consumer that re-populates them.

## Why the slow-path didn't save it

`oplib_pool_stage_to_gpu` has a "slow path" that allocates SASS via `ga10b_gmmu_alloc` when `h->shader_*` is zero. The slow path succeeds (the `oplib stage: ok, gpu_va_base=0x4000000000` log line) and `g_sass_pool_gpu_va` ends up valid. But:

1. The slow-path allocation only covers `sass_len` bytes (the SASS region), not the 256 KB (`OPLIB_POOL_MIN_BYTES`) the W-series dispatch path needs for SASS + 3 scratch slots.
2. The dispatch FFIs read `h->shader_phys + OPLIB_POOL_OFF_SCRATCH<N>` for per-call CPU↔GPU activation staging. With `h->shader_phys = 0`, that math gives an invalid CPU phys to memcpy into. The FFI's guard catches this and returns -1.

The slow path makes the SASS pool itself usable for kernel-fetch, but doesn't expose scratch slots in a form the dispatch FFI can use.

### Option A — helper-side: publish the v4 scratch region from v7

Smallest patch. Have `gpu_write_handoff_v7` (and `_v6`) allocate a single 256 KB contiguous DMA buffer alongside the per-op layout and publish its phys/gpu_va via `shader_phys`/`shader_gpu_va`/`shader_size`. The dispatch FFI's existing fast path (and `oplib_pool_stage_to_gpu`'s fast path) then "just work."

Trade-off: ties the W-series dispatch path to "MNIST helper is running, and is at a v7+ revision that allocates the v4 scratch region." Brittle if a different helper (e.g., a future pure-SLM helper) lands.

### Option B — SLM-OS side: oplib_pool allocates its own scratch slot pool

Have `oplib_pool_stage_to_gpu` separately allocate a 256 KB contiguous PMM buffer via `kbuf_alloc` (which already does chunked-VA fallback), map it into the inherited channel's GMMU via `ga10b_gmmu_alloc`, and expose accessors per slot: `oplib_pool_get_scratch_gpu_va(slot_idx)` and `_phys(slot_idx)`. The dispatch FFI then uses these accessors instead of `h->shader_phys + OFF_SCRATCH<N>`.

Trade-off: bigger SLM-OS-side change (3 files, accessor design, dispatch FFI rewrite), but decouples the W-series dispatch path from the helper's handoff version forever.

### Option C — both: helper publishes scratch when it can, SLM-OS allocates when it can't

Most robust. Fast path uses helper-published shader_* when available; slow path falls back to SLM-OS-allocated scratch. The dispatch FFI is rewritten to use accessors regardless, so the source is transparent.

## Acceptance

After the fix, a single `slm prompt 0 Hello` run on `smollm` with `gpu use inference on` + `nvgpu oplib stage`:
- `slm gpu` shows `RMSNORM ok > 0`, `GQA_ATTN ok > 0`, and ideally `fallback == 0` for those two ops.
- Per-token timing should differ from CPU-only by at least the per-dispatch overhead × number of dispatches.
- Output text may or may not differ from CPU at greedy sampling, but byte-level logits will differ (FP16 quantization noise on RMSNorm activations).

## Related

- #714 — W-series umbrella (closed at PR merge)
- #764 — GPU weights pool (this issue is the blocker preventing #764's acceptance criterion from passing)
- #692 — already-closed "GPU runs MNIST identical to CPU" (resolved separately; that issue was the MNIST v6 path which has its own working dispatcher, not this W-series FFI)
- PR #831 — the dispatch counters that surfaced this

## Reproduction (verified 2026-05-14)

1. `make kernel PLATFORM=JETSON_ORIN_NANO` from `4c6b25ab` (or any post-W7 commit; counter PR optional but recommended for visibility)
2. Deploy via `scripts/jetson-deploy-retry.sh`
3. Stream a small GGUF: `python3 slm_xload.py 192.168.4.5:2323 smollm <SmolLM2-135M.gguf>`
4. From the shell: `gpu use inference on`, `nvgpu oplib stage`, `slm launch 0 --ctx 256 --sampler greedy`, `slm prompt 0 Hello`
5. `slm gpu` — observe attempts > 0 but ok == 0 for any op with a Rust-side dispatcher

**Comment by johnjezl** [comment]
## Resolved by PR #835 (merged at e74d56b8)

The entry-guard portion of this issue — `slm_runtime_dispatch_<op>_simt` returning -1 because `h->shader_*` is zeroed by the MNIST helper's v6/v7 publish paths — is fixed. Hardware-verified after deploy:

```
[oplib] stage_cbuf: allocated 4 KB cbuf at gpu_va=0x4000040000 (phys=0x13cf61000)
[oplib] stage_to_gpu: 34560 B SASS staged at gpu_va=0x4000000000 (self-allocated 256 KB pool, base phys=0x13cf80000, scratch slots mapped)
```

The W-series dispatch FFI now gets past its early-exit guards on the toggle-respecting path.

## Downstream issue surfaced

A separate failure mode appears once the entry guards pass: `ga10b_dispatch_v7_pipeline_inline` returns -1 on every attempted dispatch. Tracking as **#834**. The fix here was a prerequisite for finding that one.

## Architectural takeaway

Option B (SLM-OS-side, oplib_pool owns its own scratch + cbuf when helper handoff is empty) was the right call: it decouples the W-series dispatch path from the MNIST helper's handoff version, so future helper revisions (or a pure-SLM helper down the road) don't re-break the W-series. The new `ga10b_gmmu_map` primitive is generally useful and will likely show up in other places (e.g., #819's deterministic-handoff work).

Closing.
